### ISSUE TITLE
Please ensure binary compatibility across CS versions

### ISSUE THREAD

**Issue by aunkrig** [checkstyle/checkstyle#7448]
In the past there were several breaks in binary compatibility, which makes life very hard for me as an extension developer (examples below). To me it appears that CHECKSTYLE and SEVNTU are always *compiled* for each released version, and tested, to ensure *source* compatibility. However when it comes to *binary* compatibility (such as when you load a CS extension), users often get mysterious error messages which are *extremely* hard to track down. The larger part of issues reported against http://unkrig.de/w/Cs-contrib.unkrig.de originate from such binary imcompatibilities. Effectively, I have to maintain a table that documents which versions of the extension are compatible with which versions of CS, which is very inconvenient for my users.

Sometimes it is possible to build workarounds (using reflection), sometimes it is not. But these workarounds "pollute" my code, and I have no criterion when I can remove each workaround -- very ugly.

My feeling is that you exercise little or no regression testing for *binary compatibility*, which, in my opinion, is absolutely crucial for a tool that claims to be extensible. Especially you should ensure that *older* versions of extensions still work with *younger* versions of CS-core, because extensions usually have a slower release cycle than the core product, and, in case of problems, it is easier for users to upgrade the core product than third-party extensions. Also, since CS is often bundled with other products (e.g. ECLIPSE), users often *cannot refuse* to upgrade the core product.

How about running your regression test suite against combinations of *older* SEVNTU versions and *younger* CS versions? That would have detected all the problems that I had in the past.

Examples for binary incompatibilities in the past:

(1)
The integer constants in `TokenTypes` have been augmented for newer Java language features (good), but, apart from that, were unnecessarily changed *several times*, breaking binary compatibility each and every time:

5.8 => 5.9
5.9 => 6.0
6.1 => 8.12 (versions in between were not analysed)

(2)
The parents of some checks (or filters) were changed, breaking each and every existing CS configuration that uses the check, and each and every tool that generates CS configurations (e.g. ECLIPSE-CS). E.g. `LineLengthCheck`:

8.23 => 8.24 Base class changed from `AbstractCheck` to `AbstractFileSetCheck` (so parent changed from `TreeWalker` to `Checker`

One ugly side effect of this change is that because `SuppressWithNearbyCommentFilter` affects only checks that or children of `TreeWalker`, it no longer suppresses messages issued by `LineLengthCheck`. The solution would be to change the superclass of `SuppressWithNearbyCommentFilter` (and all its friends) from `TreeWalkerFilter` to `Filter`. (Like I did for my `SpuuressionRegex` filter.)

(3)
The class hierarchy around `DetailAST` has changed several times:

5.9 => 6.0: Parent class changed from `CommonAST` to `CommonASTwithHiddenTokens`
8.20 => 8.21: `DetailAST` is no longer a class but an interface - creates tons of `IncompatibleClassChangeError`
8.24 => 8.25: 9 Methods were removed - again, tons of runtime errors

I mean -- `DetailAST` is *the* central API for all `TreeWalker`-based checks -- why does anyone believe it is a good idea to modify that all the time??

Summary: The continuous changes in the CS core framework meant little improvements for me (apart from the support of newer Java features), but caused lots of trouble.

Honestly, I'm considering to give up on cs-contrib, because it gives me and my users constant frustration to keep it alive.

Could you please leave the framework and APIs untouched in the future, and, rather than make incompatible changes to the existing checks and filters, deprecate them and add *new* checks and filters, e.g. `LineLengthCheck` (parent `TreeWalker`) => deprecated, and new check `LineLengthCheck2` (parent `Checker`).

**Comment by aunkrig** [comment]
Currently, ECLIPSE 201912 is unusable for me, because it comes with CS 8.27, where `SuppressWithNearbyCommentFilter` does not work for `LineLengthCheck`, and I drown in tons of `LineLength` messages.

**Comment by aunkrig** [comment]
The logic of `SuppressWithNearbyCommentFilter.Tag.isMatch()` was changed several times:
* Up to version 8.22, it was (roughly) `check || message` (where "check" means "the check id matches the checkFormat regex" and "message" means "the message matches the messageFormat regex").
* In version 8.23, the logic was changed to `(check || id) && message`
* In version 8.24, the logic was changed again to `check && id && message`

Effectively, no rule author understands any more what is going on (including me, before I examined the source code forensically). The basic problem is that three regexes are too little to implement many expressions that are useful in practice, e.g. `check && message` and `check || message`. One simple solution would be to concatenate check, id and messge, e.g. like this:

    check:de.unkrig.cscontrib.checks.NameSpelling message:Enum constant 'CONstant' entspricht nicht '^[A-Z][A-Z0-9]*(?:__?[A-Z0-9]+)*$' id:

, and then apply only *one* regex, like

    NameSpelling.*Enum constant     (implements check && message)
    NameSpelling|Enum constant      (implements check || message)

**Comment by romani** [comment]
> which makes life very hard for me as an extension developer

Please add your project to https://checkstyle.org/index.html at "Active tools". It good to know about your project existence, lets be friends and collaborate together.
The best way to collaborate it to use your extension Checks (http://cs-contrib.unkrig.de/csdoc/) in our build process, so we will run your Checks on our code on each commit. So we will know far in advance that we are going to make problem for you.
We need help to improve our CI to keep validation of compatibility.
Some breaking are done on purpose (api cleanup process), but we missed major version bump (human factor). 

>  To me it appears that CHECKSTYLE and SEVNTU are always compiled for each released version, and tested, to ensure source compatibility.

unfortunately no. We do need contributors to help us to break dependency of sevntu to eclipse-cs (that was almost dead few month ago) project - https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/705 .
We do not have time to test all, if somebody have time to make it automated ... please be welcome.


> users often get mysterious error messages which are extremely hard to track down.

please complain about such cases, or show us, we will try to help.

> I have to maintain a table that documents which versions of the extension are compatible with which versions of CS, which is very inconvenient for my users.

might be helpful also - https://checkstyle-addons.thomasjensen.com/checkstyle-compatibility-matrix.html
Such tables are unavoidable, as in version 8.X we started series of API changes. We postponed such changes for long time, there are still bunch of work to be done, to make API more reasonable.

> My feeling is that you exercise little or no regression testing for binary compatibility

almost right assumption, we have a lot of room to improve in this area. I would love for your project to appear in our CI. We can not keep all in our heads, CI knows all details of what to verify.

> you should ensure that older versions of extensions still work with younger versions of CS-core

I cannot promise this. But we definitely need to improve relationship with extensions and ease migration and .... . Our API is mess of implementations, inconsistency, hacks, bad design, .... . We will continue to cleanup API step by step.

> since CS is often bundled with other products (e.g. ECLIPSE), users often cannot refuse to upgrade the core product.

yes, there are problems in design of extensions too, they rely of core project too much (eclipse-cs, sonar). There are extensions that do not depend that much: Visual Code, IDEA.
If extension is depending on core, it should have its own strategy on releases, projects in checkstyle organization are released almost together for simplicity of maintenance (we do need more contributors to make to more smart.)

> How about running your regression test suite against combinations of older SEVNTU versions and younger CS versions?

Sure! There are even issue on this some where, we need somebody to contribute this.
Here it is - https://github.com/checkstyle/checkstyle/issues/6784
It is only recently become possible, as eclispe-cs was taken by other maintainer and it unblocks sevntu. But still sevntu have problems kin dependency. If your project do not have such dependency to eclipse-cs, we should use your project in our CI.

> (1) (2) (3)

yes, there was such changes. Some of them were done on purpose, some of them my mistake as I did not predict compatibility problems. Sorry, too much was on me only for long time, checkstyle is my hobby project, it is all my fault, this is the best I could do.

> I mean -- DetailAST is the central API for all TreeWalker-based checks -- why does anyone believe it is a good idea to modify that all the time??

Yes is it in API, .... and it is not a API, it is implementation (almost all in API package are implementations that blocks us to improve checkstyle). It is huge problem for us that historically it appeared in API. But we need to change it to put more features in checkstyle. We do not have time to make a change in one release, so It was my decision to update it when we have time. We do not block functional fixes, we release periodically, most of such decisions are based on lack of time to maintain project, so it is the best model for us.

> Summary: The continuous changes in the CS core framework meant little improvements for me (apart from the support of newer Java features), but caused lots of trouble.

for you - yes. I have the same problems with sevntu. But users (who do not use extensions) got a lot of new long awaited features.

> Could you please leave the framework and APIs untouched in the future

I can not promise this. Current API blocks us to implement Multithreading, make Checks reliable in validation ( without huge set of false positives that frustrate users and force them to disable Check), .... . We will continue to cleanup for better future.

> LineLengthCheck (parent TreeWalker) => deprecated, and new check LineLengthCheck2 (parent Checker).

This update was my fault, to allow such update without providing filter. We get to know about this only after release - https://github.com/checkstyle/checkstyle/issues/2116 .
We just need to implement support for it https://github.com/checkstyle/checkstyle/issues/7064 .

> ECLIPSE 201912 is unusable for me, because it comes with CS 8.27, where SuppressWithNearbyCommentFilter does not work for LineLengthCheck, and I drown in tons of LineLength messages.

we need https://github.com/checkstyle/checkstyle/issues/7064 to be implemented.

>  The basic problem is that three regexes are too little to implement many expressions that are useful in practice,

please be welcome to create new issue and describe a problem in details. It is better to report problem when you run into them, there might be unwanted regression from our side.

**Comment by romani** [comment]
from http://unkrig.de/w/Cs-contrib.unkrig.de
>This project uses the CS doclet (internally) to generate its CheckStyle and eclipse-cs metadata files. 

It is so long awaited feature! It would be awesome to have such functionality to help all plugins to generate metadata (it is manual process now, it is a pain for plugins and main reason why plugin development is painful).
@Calixte, @muhlba91, FYI.

**Comment by aunkrig** [comment]
Hey Romani,

> Please add your project to https://checkstyle.org/index.html at "Active tools".

Good idea! However I don't see an EDIT button, nor a way to register myself for editing. Can you please direct me?

> The best way to collaborate is to use your extension Checks (http://cs-contrib.unkrig.de/csdoc/) in our build process

Yeah... I'll look into it. I hope that we could run regression tests against *multiple* versions of CS-CONTRIB? What is the way to integrate CS-CONTRIB into the CI? A PR?

> We do need contributors to help us to break dependency of sevntu to eclipse-cs

I believe that you (the team) should *absolutely* focus on backwards compatibility, even at the price of not-so-nice software designs. CS is at a maturity stage where clean design and new features don't matter as much, but *reliability* does!! Thousands of people (or more) use it, and at most they want *careful* improvements and bug fixes.

> > users often get mysterious error messages which are extremely hard to track down.
> please complain about such cases, or show us, we will try to help.

Well, the 90% case is

com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module TreeWalker - cannot initialize module de.unkrig.cscontrib.checks.ZeroParameterSuperconstructorInvocation - Fehler beim Instanziieren der Klasse 'de.unkrig.cscontrib.checks.ZeroParameterSuperconstructorInvocation'. Es war auch nicht möglich, sie als null zu instanziieren. Bitte prüfen Sie, dass der Klassenname als kanonischer Name angegeben wurde oder lesen Sie, wie Sie abgekürzte Namen konfigurieren: https://checkstyle.org/config.html#Packages. Bitte prüfen Sie auch, dass der angegebene ClassLoader des Checkers richtig konfiguriert ist."

Happy debugging! (In this case, it was a `NoClassDefFoundError`.)

> Such tables are unavoidable, as in version 8.X we started series of API changes. We postponed such changes for long time, there are still bunch of work to be done, to make API more reasonable.

The APIs are down the drain anyway... please stop "fixing" them!!

> We will continue to cleanup API step by step.

Don't! Don't! Don't!

> projects in checkstyle organization are released almost together for simplicity

Then CS is not what can be called "extensible", but one big monolith. I would not go into that direction.

> If your project do not have such dependency to eclipse-cs, we should use your project in our CI.

CS-CONTRIB functions both independent from eclipse-cs, and in combination with eclipse-cs. (However, there have never been any problems with new versions of eclipse-cs.) I propose that it is enough to test CS-CONTRIB in combination with CS, and not in combination with eclipse-cs.

> But users (who do not use extensions) got a lot of new long awaited features.

It is good that you improve existing checks and add new ones (although I believe they should not be part of the CS core). But users don't need changes in the core API!

> > ECLIPSE 201912 is unusable for me, because it comes with CS 8.27,...

> we need #7064 to be implemented.

Unfortunately it will probably take many month until the fixes makes it into CS, then ECLIPSE-CS, then ECLIPSE...

**Comment by rnveach** [comment]
> However I don't see an EDIT button, nor a way to register myself for editing

Send a PR with a modification to https://github.com/checkstyle/checkstyle/blob/master/src/xdocs/index.xml.vm . All our pages are listed in our xdocs folder.

> What is the way to integrate CS-CONTRIB into the CI? A PR?

A PR. Basically your project can be added to https://github.com/checkstyle/checkstyle/blob/master/wercker.yml which is our current CI for running regression on external projects. You are free to send us a PR and we will guide you as needed.

Once your project is added, when another contributor does a change the CI would take the PR version of checkstyle and run it against your project looking for build failures or new violations. As mentioned above, all current regression recompiles the source in maven. We have no regressions that use compiled JARs as an example. Not saying its impossible, just saying we don't have any right now.

**Comment by romani** [comment]
> I hope that we could run regression tests against multiple versions of CS-CONTRIB?

we can run only on latest code, we do not support old versions of checkstyle at all.
If we will decide to break compatibility, your project will not stop us, but we will mostlikely provide migration.

> should absolutely focus on backwards compatibility

no, till we fix API and all old Checks, it will never be a priority. 
Even right now we have in PR update that will be breaking compatibility, and users configs will be affected. To fix a lot of issues in Check it is better to fix bad design. 

> CS is at a maturity stage where clean design and new features don't matter as much, but reliability does!!

almost 500 open issues in our issue tracker, tell me a bit different. CS is such stage that produce tons of false-positives, that force users disable Check at all.

> Then CS is not what can be called "extensible", but one big monolith.

sevntu is good example, and does not have that much problems. 
Nothing will stay as is, live is going ... if you do not want to go together, use can freeze on some old version of checkstyle and enjoy life (old versions of plugins are also available).

> However, there have never been any problems with new versions of eclipse-cs

Because there was no changes for many many years. Ones this project get good set of maintainers, it will change.

> But users don't need changes in the core API!

User do need, as with some modifications we can do better verifications and close numerous issues.
All extensions are doomed to keep-up with core, to be compatible.
the same problems have all pluging to all other applications (plugins to browsers, plugins to Sonarqube, plugin to whatever ... ), plugins stop to work with new versions very frequently.
If you do not like this, do not even start extension creation.
If you are so frustrated ...... I can only recommend you to contribute all Checks in your project to sevntu project to let your Check be alive to some extend in that project. 

> Unfortunately it will probably take many month until the fixes makes it into CS, then ECLIPSE-CS, then ECLIPSE...

if you want it quicker ... just send us PRs for all that projects and features that you need. We can do releases at any time, fact that we do this ones a month is just habit, to not spend to much time in releases. if release proccess become complete automated, we can do releases at any time we need.

**Comment by romani** [comment]
I am closing this issue, as it is discussion, and compatibility will not be restored.
All future breaking compatibility changes we will try to be accurate.

Discussion can be continued in this issue.
